[CA-946] Introducing, the MC-RBS (Resource Buffering Service)

andy7i opened the pull request:
https://broadworkbench.atlassian.net/browse/CA-946

Adds client library for RBS, with retries when we get a 500.

Config changes will be in a separate firecloud-develop PR: https://github.com/broadinstitute/firecloud-develop/pull/2354


- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

gpolumbo-broad committed:
CA-966 create a V2 workspace with a dummy googleProjectId

gpolumbo-broad committed:
CA-966 When creating a workspace, Rawls checks that it has permissions to be able to create resources, namely Google Projects, with the specified Billing Account.  Updates the BillingProject.invalidBillingAcct field while checking latest IAM on Google.  Deleted MockBillingHttpGoogleServicesDAO.scala because it appeared to be unused.  Started modifying tests so that they disregard BillingProject.status during Workspace creation because that field is irrelevant in V2 Workspaces and will be removed in https://broadworkbench.atlassian.net/browse/CA-947

gpolumbo-broad committed:
CA-966 minor refactor to new logic

gpolumbo-broad committed:
CA-966 Added test to check scenario when Google IAM check returns false

gpolumbo-broad committed:
CA-966 Getting all the wiring done for adding the Workspace's Google Project to the Billing Project's Service Perimeter during Workspace creation

gpolumbo-broad committed:
CA-966 More work for getting Service Perimeter stuff in place

gpolumbo-broad committed:
CA-966 Changing Workspace model and corresponding database table, to rename workspace.google_project -> workspace.google_project_id and add a google_project_number column

gpolumbo-broad committed:
CA-966 Making GoogleProjectNumber on Workspace model required

gpolumbo-broad committed:
CA-966 Code for updating Service Perimeter and (hopefully) polling until the operation is complete.

gpolumbo-broad committed:
CA-966 Added method for setting Billing Account on project

gpolumbo-broad committed:
CA-966 method signature fix

gpolumbo-broad committed:
CA-966 Added method for setting Billing Account on project

gpolumbo-broad committed:
CA-966 Cleaning up code, scaladocs, commenting, renaming, etc.

gpolumbo-broad committed:
CA-966 Added new query to lookup Workspaces for a Service Perimeter in a single query.  Renamed the WorkspaceQuery.save method to be createOrUpdate to more precisely describe what the method does

gpolumbo-broad committed:
CA-966 Somehow missed a bunch of files from the save -> createOrUpdate renaming

gpolumbo-broad committed:
CA-966 Scaladoc

andy7i committed:
initial - adding RBS in Rawls - needs feedback / answers

andy7i committed:
temp - make compiler happy

gpolumbo-broad committed:
CA-966 Made Workspace.google_project_number field optional in Rawls Workspace model

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/config/RbsConfig.scala:
```diff
@@ -0,0 +1,20 @@
+
+import com.typesafe.config.Config
+
+case class RbsConfig(
```
In general in this PR I think I'd prefer `ResourceBuffer` or something like it over `Rbs`

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1834,12 +1835,47 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-  private def getGoogleProjectFromRBS: Future[GoogleProjectId] = {
-    // TODO: Implement for real
-    Future.successful(GoogleProjectId(UUID.randomUUID.toString.substring(0, 8) + "_fake_proj_name"))
+  // todo: find this a home
```
Maybe `ResourceBufferModel`?

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/rbs/RbsService.scala:
```diff
@@ -0,0 +1,32 @@
+    new RbsService(rbsDao, userInfo, serviceAccountCreds)
+  }
+}
+// todo: delete this file? might not need this if we're just using RBS under the hood.
```
I think it depends on how much work needs to get done here beyond just making calls to RBS. `RbsDAO` should just be the API calls and some retry logic like you have it in my opinion. In its current state, I'd agree that we could probably use `RbsDAO` directly from `WorkspaceService`, but I could see wanting to move stuff like `generateHandoutRequestId`, `getProjectPoolId` and `getGoogleProjectFromRBS` into this class

andy7i committed:
rename rbs to resourceBuffer

gpolumbo-broad committed:
CA-966 Broken tests!  Trying to figure out Mockito

andy7i committed:
restructuring, config set up

gpolumbo-broad committed:
CA-966 Refactored WorkspaceServiceConfig to be its own file with a companion object for constructing instances from a Config

gpolumbo-broad committed:
CA-966 Making my way through tests

andy7i committed:
remove retry, Future

andy7i committed:
make use of types ProjectPoolId and PoolId

andy7i committed:
add future and retry

andy7i committed:
we aren't using getPoolInfo, so remove it

andy7i committed:
added MockResourceBufferDAO, optimize imports, move code

gpolumbo-broad committed:
CA-966 First Service Perimeter unit test

andy7i committed:
make local rawls happy

gpolumbo-broad committed:
CA-966 Service Perimeter test added that makes sure the correct list of projects is built and sent to Google to overwrite projects in perimeter

gpolumbo-broad committed:
CA-966 Cleanup on the test for setting the Billing Account on the Google Project

andy7i committed:
unit test, comments

gpolumbo-broad committed:
CA-966 More tests

andy7i committed:
rename unit test, remove todo's:

  // todo: put in firecloud-develop rawls config
  /*

  gcs {
  ...
  pathToResourceBufferPem = /etc/rbs-account.pem
  }

  resourceBuffer {
    projectPool {
      regularProjectPoolId = "regularProjectPoolId-manualentry"
      servicePerimeterProjectPoolId = "servicePerimeterProjectPoolId-manualentry"
    }
    url =
  }
   */

  // todo: rebase

andy7i committed:
post-rebase: make tests happy

andy7i committed:
grab client SA email from config

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1030,6 +1032,16 @@ class HttpGoogleServicesDAO(
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(resourceBufferPemEmail)
+      .setServiceAccountScopes(Seq(ComputeScopes.CLOUD_PLATFORM).asJava)
```
not entirely sure what scopes are needed here, but it errors if there are no scopes.

andy7i commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/resourcebuffer/ResourceBufferSpec.scala:
```diff
@@ -0,0 +1,25 @@
+  val resourceBufferService = new ResourceBufferService(resourceBufferDAO, userInfo)
+
+
+  "getGoogleProjectFromRBS" should "get a Google Project ID" in {
```
I wasn't really sure if any unit tests would be useful / worth the effort but can gladly add more if someone has more ideas

marctalbott replied:
Greg's PR has a couple of tests that `createWorkspace` and `cloneWorkspace` can get a project from RBS that could be implemented

s-rubenstein replied:
Yeah, I would rather test this inside of another test, than give it its on test. (Adding something like `verifyResourceBufferService isCalled`

andy7i commented on core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala:
```diff
@@ -15,13 +15,14 @@ import org.broadinstitute.dsde.rawls.config.{DataRepoEntityProviderConfig, Deplo
 import org.broadinstitute.dsde.rawls.coordination.UncoordinatedDataSourceAccess
 import org.broadinstitute.dsde.rawls.dataaccess._
 import org.broadinstitute.dsde.rawls.dataaccess.datarepo.DataRepoDAO
+import org.broadinstitute.dsde.rawls.dataaccess.resourcebuffer.ResourceBufferDAO
```
intellij's optimize imports did some stuff

gpolumbo-broad committed:
CA-966 Uniqueness constraint on Workspace.google_project_number column

gpolumbo-broad committed:
CA-966 Uniqueness constraint on Workspace.google_project_number column

gpolumbo-broad committed:
CA-966 Forgot an override

gpolumbo-broad committed:
CA-966 perhaps HttpGoogleServicesDAO.setBillingAccountForProject should, you know, actually set the Billing Account maybe?

gpolumbo-broad committed:
CA-966 Removing commented out code

andy7i committed:
change scope

gpolumbo-broad committed:
CA-966 PR fix, only updating BillingProject.invalidBillingAccount if the value has changed

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/config/ResourceBufferConfig.scala:
```diff
@@ -0,0 +1,20 @@
+object ResourceBufferConfig{
+  def apply(conf: Config): ResourceBufferConfig = {
+    ResourceBufferConfig(
+      conf.getString("url"),
```
Where is the firecloud-develop PR that adds these values to config?

andy7i replied:
I needed to futz around with decoding the secret, but it's here: https://github.com/broadinstitute/firecloud-develop/pull/2354

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/config/ResourceBufferConfig.scala:
```diff
@@ -0,0 +1,20 @@
+case class ResourceBufferConfig(
+                                 url: String,
+                                 regularProjectPoolId: ProjectPoolId,
+                                 servicePerimeterProjectPoolId: ProjectPoolId
```
What's different about the regular projects and the service perimeter projects? I thought that we were adding projects to the perimeter once we got them, not that they'd be in the perimeter already?

andy7i replied:
RBS gives projects with APIs enabled (for example, this is the [AoU project pool configuration](https://github.com/DataBiosphere/terra-rbs/blob/master/src/main/resources/config/dev/resource-config/aou_ws_0_0_1.yml)). The idea here was that the service perimeter could have different APIs enabled. Is this not useful/necessary to have?

marctalbott replied:
I don't have access to the link to the AoU project pool configuration, so I might be misunderstanding this.

I don't know of any difference in APIs that need to be enabled for a project to be eligible to be added to a service perimeter. That said, we do currently support a couple of small variations in projects (highSecurity, flowLogs, privateIpGoogleAccess -- all the boolean flags currently on the [createBillingProject endpoint](https://rawls.dsde-dev.broadinstitute.org/#/billing/createBillingProjectFull)). I could envision having different project pools to support those flags, but none of those flags should preclude a project from being added to a perimeter.

dvoet replied:
need to have highSecurity network to be in a perimeter. In any case, we probably want a different name than `servicePerimeterProjectPoolId` just now sure what.

andy7i replied:
Maybe this is something that we need:
```
        enableNetworkMonitoring:
          description: |-
            Make network traffic is measured. If enabled, turn on flow logs, enable private google access.
            See https://docs.google.com/document/d/1ccz2kzDL68CPofZ-b95ykQIudCgh7OCSFJ4Ym4Oz0r0/edit for more context.
```

https://github.com/DataBiosphere/terra-resource-buffer/blob/eed0f78856b9a7c7b5f6fd77f9bcfc30de63afdc/src/main/resources/config/resource_schema.yaml#L102

in which case, would something like `networkMonitoredProjectPoolId` or `flowLogsProjectPoolId` work?

dvoet replied:
how about `exfiltrationControlledPoolId`

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/ResourceBufferModel.scala:
```diff
@@ -0,0 +1,10 @@
+package org.broadinstitute.dsde.rawls.model
+
+case class ProjectPoolId(value: String)
```
What's the difference between a `PoolId` and a `ProjectPoolId`?

andy7i replied:
RBS doesn't actually have a concept of ProjectPoolId, so I separated them so that we can ensure that we aren't pulling from the wrong type of pool. Since Rawls might not be pulling any other types of resources from RBS, this might not be super useful

marctalbott replied:
In that case, I'd vote to remove `PoolId`. If we ever start using RBS for other resources, we could add distinct `{ResourceName}PoolId` case classes then if we want to ensure that we aren't pulling from the wrong type of pool. At the moment, this feels like an unnecessary distinction to me

s-rubenstein replied:
I am going to second this, this smells a bit to me of premature optimization.

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:
```diff
@@ -164,7 +165,9 @@ object Boot extends IOApp with LazyLogging {
         terraBucketWriterRole = gcsConfig.getString("terraBucketWriterRole"),
-        accessContextManagerDAO = accessContextManagerDAO
+        accessContextManagerDAO = accessContextManagerDAO,
+        resourceBufferPemFile = gcsConfig.getString("pathToResourceBufferPem"),
```
What is this service account? I would've thought that we'd be making requests to RBS as the Rawls service account?

andy7i replied:
RBS has a client SA instead of having each individual app being added to a whitelist individually. Each of the kernel apps will have a client SA like this.

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/resourcebuffer/HttpResourceBufferDAO.scala:
```diff
@@ -0,0 +1,64 @@
+    client
+  }
+
+  override def getProjectPoolId(projectPoolType: ProjectPoolType.ProjectPoolType): ProjectPoolId = {
```
Not sure that this should be part of the DAO. I think the public methods in this class should just be actual requests to RBS. This seems like it belongs in the ResourceBufferService or I could also see it as a helper function for the enum (`toProjectPoolId`)

andy7i replied:
since I would need to pass in the config, I'm not sure if I can make it part of the Enumeration, but maybe if I change it to a sealed trait?

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/resourcebuffer/ResourceBufferService.scala:
```diff
@@ -0,0 +1,37 @@
+}
+class ResourceBufferService(resourceBufferDAO: ResourceBufferDAO, protected val userInfo: UserInfo) {
+
+  def GetGoogleProjectFromRBS(projectPoolType: ProjectPoolType): Future[GoogleProjectId] = getGoogleProjectFromRBS(projectPoolType)
```
I know this is the pattern that we use elsewhere in Rawls, but I can't say I remember _why_ we use it? It's always felt unnecessary to me, but maybe that's because I just don't get it

dvoet replied:
this pattern is a remnant of the migration from spray to akka http. We don't need to propagate it in new service classes

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/resourcebuffer/ResourceBufferService.scala:
```diff
@@ -0,0 +1,37 @@
+  //  handoutRequestId:
+  //        The unique identifier presented by the client for a resource request.
+  //        Using the same handoutRequestId in the same pool would get the same resource back.
+  private def generateHandoutRequestId(userInfo: UserInfo, projectPoolId: ProjectPoolId): String = {
```
It might be worth making a case class for this too so that we aren't passing Strings around

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1030,6 +1032,16 @@ class HttpGoogleServicesDAO(
       .build()
   }
 
+  def getResourceBufferServiceAccountCredential: Credential = {
+    new GoogleCredential.Builder()
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(resourceBufferPemEmail)
+      .setServiceAccountScopes(workbenchLoginScopes.asJava) // might be ok with null scopes but adding this here per recommendation
+      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(resourceBufferPemFile))
+      .build()
+  }
```
why is this in `GoogleServicesDAO`?

andy7i replied:
It looked like all existing service account Credentials were being set up here. Do you think it makes more sense to stick it in HttpResourceBufferDAO?

marctalbott commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1860,8 +1856,10 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       case _ => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Billing Account is missing or invalid for Billing Project: ${billingProject}"))
     }
 
+    val resourceBufferService = new ResourceBufferService(resourceBufferDAO, userInfo)
```
Is this how we create services elsewhere? I'd prefer to see it get created in the constructor rather than down here

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/resourcebuffer/HttpResourceBufferDAO.scala:
```diff
@@ -0,0 +1,64 @@
+  override def handoutGoogleProject(poolId: PoolId, handoutRequestId: String): Future[GoogleProjectId] = {
+    retry(when500) { () =>
+      Future {
+        val resource = handoutResourceGeneric(poolId, handoutRequestId, OAuth2BearerToken(clientServiceAccountCreds.getAccessToken))
```
I think you need a call to `refreshToken()` before call `getAccessToken`

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/resourcebuffer/HttpResourceBufferDAO.scala:
```diff
@@ -0,0 +1,64 @@
+
+  protected def when500(throwable: Throwable): Boolean = {
+    throwable match {
+      case t: RawlsExceptionWithErrorReport => t.errorReport.statusCode.exists(_.intValue / 100 == 5)
```
I don't think RBS will throw a `RawlsExceptionWithErrorReport`

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/resourcebuffer/ResourceBufferService.scala:
```diff
@@ -0,0 +1,37 @@
+    new ResourceBufferService(resourceBufferDAO, userInfo)
+  }
+}
+class ResourceBufferService(resourceBufferDAO: ResourceBufferDAO, protected val userInfo: UserInfo) {
```
do we need a service class or can `ResourceBufferDAO` just be called from `WorkspaceService` where it is needed

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/resourcebuffer/ResourceBufferService.scala:
```diff
@@ -0,0 +1,37 @@
+
+    val projectPoolId: ProjectPoolId = resourceBufferDAO.getProjectPoolId(projectPoolType)
+
+    val handoutRequestId = generateHandoutRequestId(userInfo, projectPoolId)
```
we should not be generating random request ids. We should use the workspace id so that if we end up retrying for the same workspace we get the same project and don't leak projects

gpolumbo-broad committed:
CA-966 Cloning a Workspace should also add it to a service perimeter as needed.  Copy/pasta tests for Clone Workspace

gpolumbo-broad committed:
CA-966 Removing some commented out imports

gpolumbo-broad committed:
CA-966 Added googleProjectNumber to WorkspaceDetails in Swagger

gpolumbo-broad committed:
Merge branch 'v2_workspaces' into gpolumbo_CA_966_create_patch_v2_workspaces

# Conflicts:
#	core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/MockBillingHttpGoogleServicesDAO.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala

gpolumbo-broad committed:
CA-966 Merging latest changes from base v2_workspaces branch

andy7i committed:
Addressing PR feedback

andy7i committed:
Merge branch 'gpolumbo_CA_966_create_patch_v2_workspaces' into aj--rbs--ca-946

# Conflicts:
#	core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala
#	core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

gpolumbo-broad committed:
CA-966 Fixing tests

gpolumbo-broad committed:
CA-966 Fixing tests

andy7i committed:
fixes tests (add test config)

(will require another rebase for the fixes in Greg's branch)

andy7i committed:
Merge branch 'gpolumbo_CA_966_create_patch_v2_workspaces' into aj--rbs--ca-946

andy7i committed:
fixes tests (not sure why the Mock issue is only showing up now)

gpolumbo-broad committed:
PR feedback fix to send properly formatted billing account name to google

gpolumbo-broad committed:
Fixing WorkflowSubmissionSpec

andy7i committed:
update rbs client v

andy7i committed:
Merge branch 'gpolumbo_CA_966_create_patch_v2_workspaces' into aj--rbs--ca-946

andy7i committed:
update rbs client pt 2

andy7i commented:
I'm a bit worried about the interchangeable naming of RBS (aka Resource Buffering Service / ResourceBuffer / Buffer / MC RBS), some of which is definitely on me.

s-rubenstein commented on core/src/test/resources/reference.conf:
```diff
@@ -98,6 +98,14 @@ dataRepo {
+    regular = "dev-project-regular-v1"
+    servicePerimeter = "dev-project-serviceperimeter-v1"
+  }
+  url = "https://rbs.dsde-dev.broadinstitute.orgLOCAL"
```
Should this have a `:` between `.org` and `LOCAL`?

andy7i replied:
I think this file is used for unit tests

s-rubenstein commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/resourcebuffer/ResourceBufferSpec.scala:
```diff
@@ -0,0 +1,29 @@
+class ResourceBufferSpec extends AsyncFlatSpec {
+
+
+  val userInfo = UserInfo(RawlsUserEmail("owner-access"), OAuth2BearerToken("token"), 123, RawlsUserSubjectId("123456789876543212345"))
```
I am assuming the `RawlsUserSubjectId` is not important here? Could we have it generate a random number string instead? (Just to show that it is irrelevant?)

s-rubenstein reviewed (commented):
This all looks reasonable to me. Mainly seconding Marc's comments with a few notes.

andy7i committed:
remove poolId

andy7i committed:
PR feedback

andy7i committed:
PR feedback - move RBS unit test

andy7i committed:
PR feedback - remove unit test

tested that the mocked RBS is validated through other tests like createWorkspace

( changing this line to an empty string GoogleProjectId makes this test fail:
```    val googleProjectId = GoogleProjectId("")```

andy7i committed:
PR feedback - rename

andy7i committed:
PR feedback - rename servicePerimeter project pool to exfiltrationControlled

s-rubenstein reviewed (approved):
LGTM

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -124,7 +124,9 @@ class HttpGoogleServicesDAO(
                              terraBucketWriterRole: String,
-                             override val accessContextManagerDAO: AccessContextManagerDAO)(implicit val system: ActorSystem, val materializer: Materializer, implicit val executionContext: ExecutionContext, implicit val cs: ContextShift[IO], implicit val timer: Timer[IO]) extends GoogleServicesDAO(groupsPrefix) with FutureSupport with GoogleUtilities {
+                             override val accessContextManagerDAO: AccessContextManagerDAO,
+                             resourceBufferPemFile: String,
```
we should use the json instead of pem. That makes the firecloud-develop stuff simpler. Use `GoogleCredential.fromStream(...).toBuilder.setServiceAccountScopes(...).build()`

andy7i committed:
Merge branch 'v2_workspaces' into aj--rbs--ca-946

# Conflicts:
#	core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changelog.xml
#	core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala
#	core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/ApiServiceSpec.scala
#	core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala

andy7i committed:
fix merge fallout

andy7i committed:
remove use of RBS in favor of Buffer

andy7i committed:
get RBS SA creds from JSON instead of PEM. Removes now useless param `resourceBufferPemEmail`

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1049,6 +1050,12 @@ class HttpGoogleServicesDAO(
       .build()
   }
 
+  def getResourceBufferServiceAccountCredential: Credential = {
```
Barring how long a built `GoogleCredential` is valid for, I'm wondering if we can do all of this just one time during `Boot`.  Tests might want to use different JSON files, but the running App will only ever get 1 credential file at startup, so we have no need to keep making this API call to Google, unless the credential expires, in which case we might be able to keep track of the age of the credential and request it only when necessary to avoid extra calls to Google.

dvoet replied:
can change this to a `lazy val`

andy7i replied:
changing to a lazy val

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1857,16 +1850,21 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     * @return Future[(GoogleProjectId, GoogleProjectNumber)] of the project that we claimed from RBS
     */
-  private def setupGoogleProject(billingProject: RawlsBillingProject, span: Span = null): Future[(GoogleProjectId, GoogleProjectNumber)] = {
+  private def setupGoogleProject(billingProject: RawlsBillingProject, span: Span = null, workspaceId: String) = {
```
Keep the return type

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2105,7 +2103,7 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
 
             traceDBIOWithParent("getPolicySyncStatus", s1)(_ => DBIO.from(samDAO.getPolicySyncStatus(SamResourceTypeNames.billingProject, workspaceRequest.namespace, SamBillingProjectPolicyNames.owner, userInfo).map(_.email))).flatMap { projectOwnerPolicyEmail =>
-              traceDBIOWithParent("setupGoogleProject", s1)(_ => DBIO.from(setupGoogleProject(billingProject, s1))).flatMap { case (googleProjectId, googleProjectNumber) =>
+              traceDBIOWithParent("setupGoogleProject", s1)(_ => DBIO.from(setupGoogleProject(billingProject, s1, workspaceId))).flatMap { case (googleProjectId, googleProjectNumber) =>
```
Should we consider using `WorkspaceName` as the `handoutRequestId` here?  The `workspaceId` is a UUID we just generated here and there isn't much to tie it back to the original request.  For example, if the `saveNewWorkspace` action should fail on the next line, we won't know the `workspaceId` except for maybe/hopefully in the logs it will be in the failed SQL statement or something.  If we use `workspaceRequest.toWorkspaceName` here, it will hopefully make it easier to track/correlate the request.  

Oh, and I guess all of this is on the assumption that the purpose of the `handoutRequestId` is purely for tracking/auditing/troubleshooting requests for Resources from RBS

dvoet replied:
`handoutRequestId` is not at all for tracking/auditing/troubleshooting, it is functional. RBS will always give back the same google project for a given `handoutRequestId`. It is a way to avoid dropping resources on the floor in the case a retries.

So the code is right as is. Consider the case that a user creates a workspace, deletes it, then creates it again. We want to be sure RBS gives us back a different project.

gpolumbo-broad reviewed (commented):
Could also tests to `WorkspaceServiceSpec`.  There are a few `pending` tests in there related to getting the Project from RBS for Create and Clone workspace.  I think they can just do like some of the other tests and use Mockito to check that when we create/clone a workspace, there is exactly 1 call to `ResourceBufferService.getGoogleProjectFromBuffer` or to `ResourceBufferDAO.handoutGoogleProject` (either is probably fine, and I'd say choose whichever one is easier to Spy on.  There should be some existing tests doing the same thing for `MockGoogleServicesDAO` you can copy)

andy7i committed:
PR feedback - lazy val && disappeared return type

andy7i committed:
fill in pending tests

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -150,7 +150,7 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
   def getBucketServiceAccountCredential: Credential
 
-  def getResourceBufferServiceAccountCredential: Credential
+  val getResourceBufferServiceAccountCredential: Credential
```
can you leave this as def?

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -1303,7 +1304,14 @@ class WorkspaceServiceSpec extends AnyFlatSpec with ScalatestRouteTest with Matc
+
+    val workspace = Await.result(services.workspaceService.createWorkspace(workspaceRequest), Duration.Inf)
+
+    verify(services.resourceBufferService, Mockito.atLeastOnce()).getGoogleProjectFromBuffer(any[ProjectPoolType], any[String])
```
```suggestion
    verify(services.resourceBufferService).getGoogleProjectFromBuffer(any[ProjectPoolType], any[String])
```
Leaving the second parameter blank means it will check that the method was called exactly once. 

Of course same goes for the `cloneWorkspace` test

andy7i replied:
cool, I see that now. TIL!

andy7i committed:
PR feedback - an abstract `def` can be implemented by a `lazy val` ; mockito `verify` will verify that the method is called exactly once.
